### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jankeromnes**
Hello! I refactored a few things because copy/pasta made me sad. Also suggested some (in my opinion) less confusing names.

@espadrine let me know what you think!

### CODE REVIEW

File: lib/camp.js [hunk 1]
```diff
@@ -112,7 +112,7 @@ function jsonFromQuery(query, obj) {
 // This function updates ask.query with:
 // - application/x-www-form-urlencoded
 // - multipart/form-data
-function getQueries(ask, end) {
+function parseQuery(ask, end) {
```
Comment (jankeromnes) [inline lib/camp.js]: In the comment, you say "parse the query", and there seems to ever only be one "query".

**Comment by jankeromnes** [comment]
What I did:

- Created "generic routing unit" so that both `routeUnit` and `notfoundUnit` could use it.
- Folded `catchpath` into it, because it was the only remaining callsite, and the name was confusing (I thought it was another unit but with different params).
- Extracted the deflate/gzip code into a `compressed(ask)` function which returns a stream.

File: lib/camp.js (hunk 1 above)
Comment (espadrine) [inline lib/camp.js]: There's the x-www-form-urlencoded query and the multipart query.

File: lib/camp.js [hunk 2]
```diff
@@ -492,149 +506,103 @@ function staticUnit (server) {
...
-        options.reader = options.reader || templateReader;
-        if (!ask.res.getHeader('Content-Type')) {  // Allow overriding.
-          ask.mime(mime[p.extname(options.template).slice(1)] || 'text/plain');
-        }
+    if (filteredRoutes.length > 1) {
+      var warnMsg = 'More than one template path match' + ask.path + ':\n';
+      filteredRoutes.forEach (function (path) { warnMsg += '- ' + path[0]; });
+      log(warnMsg, 'warn');
+    }
+    var pathmatch = ask.path.match (RegExp (filteredRoutes[0][0]));
```
Comment (jankeromnes) [inline lib/camp.js]: Whoops (I changed the route structure to `{pattern, handler}` object).

File: lib/camp.js [hunk 3]
```diff
@@ -492,149 +506,103 @@ function staticUnit (server) {
...
-          ask.res.end('404\n');
-        });
+    parseQuery(ask, function gotQuery (err) {
+      if (err !== null) {
+        log('While getting queries for ' + ask.uri + ':\n'
+          + err
+          , 'error');
+      } else {
+        // params: template parameters (JSON-serializable).
+        filteredRoutes[0][1](ask.query, pathmatch, function end (params, options) {
```
Comment (jankeromnes) [inline lib/camp.js]: Whoops.

File: lib/camp.js [hunk 4]
```diff
@@ -492,149 +506,103 @@ function staticUnit (server) {
...
-      platepaths[0][1](ask.query, pathmatch, function end (params, options) {
-        options = options || {};
-        options.template = options.template || pathmatch[0];
-        options.reader = options.reader || templateReader;
-        if (!ask.res.getHeader('Content-Type')) {  // Allow overriding.
-          ask.mime(mime[p.extname(options.template).slice(1)] || 'text/plain');
-        }
+    if (filteredRoutes.length > 1) {
+      var warnMsg = 'More than one template path match' + ask.path + ':\n';
+      filteredRoutes.forEach (function (path) { warnMsg += '- ' + path[0]; });
```
Comment (jankeromnes) [inline lib/camp.js]: Whoops.

File: lib/camp.js (hunk 1 above)
Comment (jankeromnes) [inline lib/camp.js]: Isn't it either x-www-form-urlencoded data or multipart data?

Comment (espadrine) [inline lib/camp.js]: Nah, they override one another.

File: lib/camp.js [hunk 5]
```diff
@@ -484,157 +500,110 @@ function staticUnit (server) {
...
-function routeUnit (server) {
-  var templates = [];
-
-  function route (paths, literalCall) {
-    templates.push([RegExp(paths).source, literalCall]);
-  }
-
-  server.route = route;
+// Generic routing unit.
+function routingUnit (server, name) {
```
Comment (espadrine) [inline lib/camp.js]: `routingEngine`

**Comment by jankeromnes** [comment]
Addressed some nits, but let's take smaller steps instead of compulsively refactoring everything. I'll open smaller separate pull requests which will be safer to review and accept.

**Pull request closed without merging**